plugins/redis/redis.cpp

```diff
@@ -375,10 +473,250 @@ void Connection::parseOptions(ICodeContext * ctx, const char * _options)
+{
+    if (lockedKey)
+        buffer.append(REDIS_LOCK_PREFIX).append("_").append(keyOrChannel).append("_").append(_database);
+    else
+        buffer.set(keyOrChannel);
+    return buffer.str();
+}
+bool SubConnection::isCorrectChannel(const redisReply * reply, const char * op) const
+{
+    if (!reply || channel.isEmpty())
+        return false;
+    return (reply->type == REDIS_REPLY_ARRAY)\
+            && (reply->elements > 1)\
+            && (strcmp(op, reply->element[0]->str) == 0)\
+            && (strcmp(channel.str(), reply->element[1]->str) == 0);
+}
+void SubConnection::subscribe(ICodeContext * ctx, const char * _channel)
+{
+    assert(channel.isEmpty());//If this is a reused subConnection then this implies that it was not correctly unsubscribed/reset.
+    channel.set(_channel);
+
+    OwnedReply reply = Reply::createReply(redisCommand("SUBSCRIBE %b", channel.str(), (size_t)channel.length()));
+    /* It is possible for the subscription to succeed in being registered by the server but fails in receiving the associated reply.
+     * Set early to prevent not unsubscribing when failing.
+     */
+    subscribed = true;
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+    if (!isCorrectChannel(reply->query(), "subscribe"))
+        fail("SUBSCRIBE", "failed to register SUB", channel.str());
+}
+void SubConnection::subAndWaitForSinglePub(ICodeContext * ctx, const char * keyOrChannel, size_t & messageSize, char * & message, int _database, bool lockedKey)
+{
+    StringBuffer channelBuffer;
+    encodeChannel(channelBuffer, keyOrChannel, _database, lockedKey);
+    subscribe(ctx, channelBuffer.str());
+
+    //Now wait for published message
+    OwnedReply reply = new Reply();
+    readReply(reply);
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+
+    if (isCorrectChannel(reply->query(), "message"))
+    {
+        if (reply->query()->element[2]->len > 0)
+        {
+            messageSize = (size_t)reply->query()->element[2]->len;
+            message = reinterpret_cast<char*>(allocateAndCopy(reply->query()->element[2]->str, messageSize));
+        }
+        else
+        {
+            messageSize = 0;
+            message = nullptr;
+        }
+        unsubscribe(ctx, nullptr);
+        return;
+    }
+    throwUnexpected();
+}
+int Connection::writeBufferToSocket()
+{
+    int done = 0;
+    void * aux = nullptr;
+    if (context->flags & REDIS_BLOCK)
+    {
+       /* Write until done */
+       while (!done)
+       {
+           if (redisBufferWrite(context, &done) == REDIS_ERR)
+               return REDIS_ERR;
+       }
+    }
+    return REDIS_OK;
+}
+unsigned SubConnection::getUnsubscribeReadAttempts()
+{
+    if (!unsubscribeReadAttemptsChecked)//unsubscribeReadAttemptsChecked is std:atomic<bool>. Test to guard against unnecessary critical section
```

Comment:
minor: this is overkill.  You should only use this pattern if it really is time critical (i.e. you will notice if it takes 10ns longer).  Otherwise simpler code that only checks a standard boolean inside a critical section is better.